base/server/python/pki/server/cli/migrate.py

```diff
@@ -601,3 +602,45 @@ def create_link(self, instance, source, dest):
 
         os.symlink(source, dest)
         os.lchown(dest, instance.uid, instance.gid)
+
+    def migrate_service(self, instance):
+        self.migrate_service_java_home(instance)
+
+    def migrate_service_java_home(self, instance):
+        # When JAVA_HOME in the Tomcat service config differs from the
+        # value in /usr/share/pki/etc/pki.conf, update the value in
+        # the service config.
+
+        if "JAVA_HOME" not in os.environ or not os.environ["JAVA_HOME"]:
+            logger.debug("Refusing to migrate JAVA_HOME with missing environment variable")
+
+        java_home = os.environ["JAVA_HOME"]
+        comment = '# Disabling JAVA_HOME; to override, set in /etc/pki/pki.conf\n'
+
+        # Update in /etc/sysconfig/<instance>
+        result = self.update_java_home_in_config(instance.service_conf, comment)
+        self.write_config(instance.service_conf, result)
+
+        # Update in /etc/pki/<instance>/tomcat.conf
+        result = self.update_java_home_in_config(instance.tomcat_conf, comment)
+        self.write_config(instance.tomcat_conf, result)
+
+    def update_java_home_in_config(self, path, comment):
+        result = []
+
+        with open(path, 'r') as conf_fp:
+            for line in conf_fp:
+                target = "JAVA_HOME="
+                if not line.startswith(target):
+                    result.append(line)
+                else:
+                    new_line = '#' + line
```

Comment:
If we're removing JAVA_HOME, IIUC we will need to create an upgrade script to update the service files for existing instances. The changes to service files below will only affect new instances. Or we can deal with the removal some other time.

Reply:
Not for systemd service files. You're right that we need to remove it from `tomcat.conf` & c for new instances, I'll fix that.

Reply:
I understand now. Yeah, perhaps we do this later when we consolidate stuff, hmmm.